README.md
# uas-parser

[![Build Status](https://travis-ci.org/GUI/uas-parser.png)](https://travis-ci.org/GUI/uas-parser)

A user agent string parser for Node.js. Returns detailed user agent information from the [user-agent-string.info](http://user-agent-string.info) project.

The internal data used for parsing is automatically updated on weekly basis from [user-agent-string.info's](http://user-agent-string.info) latest downloads.

## Usage

### parse(userAgent)

Parse the given user agent string. Returns an object containing browser details and operating system information.

#### Examples

```js
var uaParser = require('uas-parser');

uaParser.parse('Mozilla/5.0 (Macintosh; Intel Mac OS X 10_7_5) AppleWebKit/536.26.17 (KHTML, like Gecko) Version/6.0.2 Safari/536.26.17');
// { type: 'Browser',
//   uaFamily: 'Safari',
//   uaName: 'Safari 6.0.2',
//   uaUrl: 'http://en.wikipedia.org/wiki/Safari_%28web_browser%29',
//   uaCompany: 'Apple Inc.',
//   uaCompanyUrl: 'http://www.apple.com/',
//   uaIcon: 'http://user-agent-string.info/pub/img/ua/safari.png',
//   uaInfoUrl: 'http://user-agent-string.info/list-of-ua/browser-detail?browser=Safari',
//   osFamily: 'OS X',
//   osName: 'OS X 10.7 Lion',
//   osUrl: 'http://www.apple.com/osx/',
//   osCompany: 'Apple Computer, Inc.',
//   osCompanyUrl: 'http://www.apple.com/',
//   osIcon: 'http://user-agent-string.info/pub/img/os/macosx.png',
//   deviceType: 'Personal computer',
//   deviceIcon: 'http://user-agent-string.info/pub/img/device/desktop.png',
//   deviceInfoUrl: 'http://user-agent-string.info/list-of-ua/device-detail?device=Personal computer' }

uaParser.parse('Mozilla/5.0 (compatible; bingbot/2.0; +http://www.bing.com/bingbot.htm)');
// { type: 'Robot',
//   uaFamily: 'bingbot',
//   uaName: 'bingbot/2.0',
//   uaUrl: 'http://www.bing.com/webmaster/help/which-crawlers-does-bing-use-8c184ec0',
//   uaCompany: 'Microsoft Corporation',
//   uaCompanyUrl: 'http://www.microsoft.com/',
//   uaIcon: 'http://user-agent-string.info/pub/img/ua/bot_msnbot.png',
//   uaInfoUrl: 'http://user-agent-string.info/list-of-ua/bot-detail?bot=bingbot',
//   osFamily: 'unknown',
//   osName: 'unknown',
//   osUrl: 'unknown',
//   osCompany: 'unknown',
//   osCompanyUrl: 'unknown',
//   osIcon: 'http://user-agent-string.info/pub/img/os/unknown.png',
//   deviceType: 'Other',
//   deviceIcon: 'http://user-agent-string.info/pub/img/device/other.png',
//   deviceInfoUrl: 'http://user-agent-string.info/list-of-ua/device-detail?device=Other' }

uaParser.parse('EventMachine HttpClient');
// { type: 'Library',
//   uaFamily: 'EventMachine',
//   uaName: 'EventMachine',
//   uaUrl: 'http://rubyeventmachine.com/',
//   uaCompany: '',
//   uaCompanyUrl: '',
//   uaIcon: 'http://user-agent-string.info/pub/img/ua/DLLicon.png',
//   uaInfoUrl: 'http://user-agent-string.info/list-of-ua/browser-detail?browser=EventMachine',
//   osFamily: 'unknown',
//   osName: 'unknown',
//   osUrl: 'unknown',
//   osCompany: 'unknown',
//   osCompanyUrl: 'unknown',
//   osIcon: 'http://user-agent-string.info/pub/img/os/unknown.png',
//   deviceType: 'Other',
//   deviceIcon: 'http://user-agent-string.info/pub/img/device/other.png',
//   deviceInfoUrl: 'http://user-agent-string.info/list-of-ua/device-detail?device=Other' }
```

### lookup(userAgent)

Provides the same functionality as `parse(userAgent)`, but caches the results for the given user agent string in memory. This can provide faster lookups when repeatedly parsing identical user agent strings.

The last 5,000 user agent strings seen will be cached. Cached results will remain valid for up to 1 week.

## Other User Agent Parsers

- [useragent](https://npmjs.org/package/useragent): Derived from [browserscope.org's](http://www.browserscope.org/) user agent parser. An excellent user agent parser for web browsers, but I found user-agent-string.info's data better at categorizing and dealing with robots, software libraries, and other more esoteric user agents (at least for my use case).

## Credits

All user agent data is from [user-agent-string.info](http://user-agent-string.info). Data is licensed under a [Creative Commons Attribution 3.0 Unported License](http://creativecommons.org/licenses/by/3.0/deed.en_US).


CHANGELOG.md
# uas-parser Change Log

## 0.2.3 - 2018-03-04

### Security
- Prevent regex denial of service attacks by only parsing the first 1,000 characters of the user agent string. Thanks to [@davisjam](https://github.com/davisjam). ([#5](https://github.com/GUI/uas-parser/pull/5))

## 0.2.2 - 2014-10-03

### Fixed
- Fix so this module doesn't keep node processes alive forever.

## 0.2.1 - 2014-09-19

### Security
- Upgrade "request" dependency to 2.44.0 (see [#3](https://github.com/GUI/uas-parser/pull/3)).

## 0.2.0 - 2013-10-27

### Added
- Return new devices information from user-agent-string.info. This
  categorizes the type of device for requests (PCs, tablets, phones, TVs, etc).

## 0.1.1 - 2013-08-02

### Fixed
-  Allow only 1 process to perform the automatic data update task
  concurrently. Helpful when used inside a cluster to prevent multiple
  processes from all trying to download the data simultaneously.

## 0.1.0 / 2013-06-16

- Initial release


test/parsing.js
'use strict';

require('chai').should();

var uasParser = require('../');

describe('parse', function() {
  it('parses a browser', function() {
    var ua = uasParser.parse('Mozilla/5.0 (Macintosh; Intel Mac OS X 10_7_5) AppleWebKit/536.26.17 (KHTML, like Gecko) Version/6.0.2 Safari/536.26.17');

    Object.keys(ua).length.should.eql(17);
    ua.type.should.eql('Browser');
    ua.uaFamily.should.eql('Safari');
    ua.uaName.should.eql('Safari 6.0.2');
    ua.uaUrl.should.eql('http://en.wikipedia.org/wiki/Safari_%28web_browser%29');
    ua.uaCompany.should.eql('Apple Inc.');
    ua.uaCompanyUrl.should.eql('http://www.apple.com/');
    ua.uaIcon.should.eql('http://user-agent-string.info/pub/img/ua/safari.png');
    ua.uaInfoUrl.should.eql('http://user-agent-string.info/list-of-ua/browser-detail?browser=Safari');
    ua.osFamily.should.eql('OS X');
    ua.osName.should.eql('OS X 10.7 Lion');
    ua.osUrl.should.eql('http://www.apple.com/osx/');
    ua.osCompany.should.eql('Apple Computer, Inc.');
    ua.osCompanyUrl.should.eql('http://www.apple.com/');
    ua.osIcon.should.eql('http://user-agent-string.info/pub/img/os/macosx.png');
    ua.deviceType.should.eql('Personal computer');
    ua.deviceIcon.should.eql('http://user-agent-string.info/pub/img/device/desktop.png');
    ua.deviceInfoUrl.should.eql('http://user-agent-string.info/list-of-ua/device-detail?device=Personal computer');
  });

  it('parses a robot', function() {
    var ua = uasParser.parse('Mozilla/5.0 (compatible; bingbot/2.0; +http://www.bing.com/bingbot.htm)');

    Object.keys(ua).length.should.eql(17);
    ua.type.should.eql('Robot');
    ua.uaFamily.should.eql('bingbot');
    ua.uaName.should.eql('bingbot/2.0');
    ua.uaUrl.should.eql('http://www.bing.com/webmaster/help/which-crawlers-does-bing-use-8c184ec0');
    ua.uaCompany.should.eql('Microsoft Corporation');
    ua.uaCompanyUrl.should.eql('http://www.microsoft.com/');
    ua.uaIcon.should.eql('http://user-agent-string.info/pub/img/ua/bot_msnbot.png');
    ua.uaInfoUrl.should.eql('http://user-agent-string.info/list-of-ua/bot-detail?bot=bingbot');
    ua.osFamily.should.eql('unknown');
    ua.osName.should.eql('unknown');
    ua.osUrl.should.eql('unknown');
    ua.osCompany.should.eql('unknown');
    ua.osCompanyUrl.should.eql('unknown');
    ua.osIcon.should.eql('http://user-agent-string.info/pub/img/os/unknown.png');
    ua.deviceType.should.eql('Other');
    ua.deviceIcon.should.eql('http://user-agent-string.info/pub/img/device/other.png');
    ua.deviceInfoUrl.should.eql('http://user-agent-string.info/list-of-ua/device-detail?device=Other');
  });

  it('parses a library', function() {
    var ua = uasParser.parse('curl/7.30.0');

    Object.keys(ua).length.should.eql(17);
    ua.type.should.eql('Library');
    ua.uaFamily.should.eql('cURL');
    ua.uaName.should.eql('cURL 7.30.0');
    ua.uaUrl.should.eql('http://curl.haxx.se/');
    ua.uaCompany.should.eql('team Haxx');
    ua.uaCompanyUrl.should.eql('http://www.haxx.se/');
    ua.uaIcon.should.eql('http://user-agent-string.info/pub/img/ua/curl.png');
    ua.uaInfoUrl.should.eql('http://user-agent-string.info/list-of-ua/browser-detail?browser=cURL');
    ua.osFamily.should.eql('unknown');
    ua.osName.should.eql('unknown');
    ua.osUrl.should.eql('unknown');
    ua.osCompany.should.eql('unknown');
    ua.osCompanyUrl.should.eql('unknown');
    ua.osIcon.should.eql('http://user-agent-string.info/pub/img/os/unknown.png');
    ua.deviceType.should.eql('Other');
    ua.deviceIcon.should.eql('http://user-agent-string.info/pub/img/device/other.png');
    ua.deviceInfoUrl.should.eql('http://user-agent-string.info/list-of-ua/device-detail?device=Other');
  });

  it('parses an agent without a version number', function() {
    var ua = uasParser.parse('EventMachine HttpClient');

    Object.keys(ua).length.should.eql(17);
    ua.type.should.eql('Library');
    ua.uaFamily.should.eql('EventMachine');
    ua.uaName.should.eql('EventMachine');
    ua.uaUrl.should.eql('http://rubyeventmachine.com/');
    ua.uaCompany.should.eql('');
    ua.uaCompanyUrl.should.eql('');
    ua.uaIcon.should.eql('http://user-agent-string.info/pub/img/ua/DLLicon.png');
    ua.uaInfoUrl.should.eql('http://user-agent-string.info/list-of-ua/browser-detail?browser=EventMachine');
    ua.osFamily.should.eql('unknown');
    ua.osName.should.eql('unknown');
    ua.osUrl.should.eql('unknown');
    ua.osCompany.should.eql('unknown');
    ua.osCompanyUrl.should.eql('unknown');
    ua.osIcon.should.eql('http://user-agent-string.info/pub/img/os/unknown.png');
    ua.deviceType.should.eql('Other');
    ua.deviceIcon.should.eql('http://user-agent-string.info/pub/img/device/other.png');
    ua.deviceInfoUrl.should.eql('http://user-agent-string.info/list-of-ua/device-detail?device=Other');
  });

  it('parses an agent with a hardcoded OS', function() {
    var ua = uasParser.parse('gvfs/1.3.1');

    Object.keys(ua).length.should.eql(17);
    ua.type.should.eql('Other');
    ua.uaFamily.should.eql('GnomeVFS');
    ua.uaName.should.eql('GnomeVFS 1.3.1');
    ua.uaUrl.should.eql('http://developer.gnome.org/doc/API/2.0/gnome-vfs-2');
    ua.uaCompany.should.eql('The GNOME Project');
    ua.uaCompanyUrl.should.eql('http://www.gnome.org/');
    ua.uaIcon.should.eql('http://user-agent-string.info/pub/img/ua/webdav.png');
    ua.uaInfoUrl.should.eql('http://user-agent-string.info/list-of-ua/browser-detail?browser=GnomeVFS');
    ua.osFamily.should.eql('Linux');
    ua.osName.should.eql('Linux');
    ua.osUrl.should.eql('http://en.wikipedia.org/wiki/Linux');
    ua.osCompany.should.eql('');
    ua.osCompanyUrl.should.eql('');
    ua.osIcon.should.eql('http://user-agent-string.info/pub/img/os/linux.png');
    ua.deviceType.should.eql('Other');
    ua.deviceIcon.should.eql('http://user-agent-string.info/pub/img/device/other.png');
    ua.deviceInfoUrl.should.eql('http://user-agent-string.info/list-of-ua/device-detail?device=Other');
  });

  it('parses an unknown user agent', function() {
    var ua = uasParser.parse('gobbledygook');

    Object.keys(ua).length.should.eql(17);
    ua.type.should.eql('unknown');
    ua.uaFamily.should.eql('unknown');
    ua.uaName.should.eql('unknown');
    ua.uaUrl.should.eql('unknown');
    ua.uaCompany.should.eql('unknown');
    ua.uaCompanyUrl.should.eql('unknown');
    ua.uaIcon.should.eql('http://user-agent-string.info/pub/img/ua/unknown.png');
    ua.uaInfoUrl.should.eql('unknown');
    ua.osFamily.should.eql('unknown');
    ua.osName.should.eql('unknown');
    ua.osUrl.should.eql('unknown');
    ua.osCompany.should.eql('unknown');
    ua.osCompanyUrl.should.eql('unknown');
    ua.osIcon.should.eql('http://user-agent-string.info/pub/img/os/unknown.png');
    ua.deviceType.should.eql('Personal computer');
    ua.deviceIcon.should.eql('http://user-agent-string.info/pub/img/device/desktop.png');
    ua.deviceInfoUrl.should.eql('http://user-agent-string.info/list-of-ua/device-detail?device=Personal computer');
  });

  it('parses long user agents', function() {
    var ua = uasParser.parse('Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0; YPC 3.2.0; SearchSystem6829992239; SearchSystem9616306563; SearchSystem6017393645; SearchSystem5219240075; SearchSystem2768350104; SearchSystem6919669052; SearchSystem1986739074; SearchSystem1555480186; SearchSystem3376893470; SearchSystem9530642569; SearchSystem4877790286; SearchSystem8104932799; SearchSystem2313134663; SearchSystem1545325372; SearchSystem7742471461; SearchSystem9092363703; SearchSystem6992236221; SearchSystem3507700306; SearchSystem1129983453; SearchSystem1077927937; SearchSystem2297142691; SearchSystem7813572891; SearchSystem5668754497; SearchSystem6220295595; SearchSystem4157940963; SearchSystem7656671655; SearchSystem2865656762; SearchSystem6520604676; SearchSystem4960161466; .NET CLR 1.1.4322; .NET CLR 2.0.50727; Hotbar 10.2.232.0; SearchSystem9616306563; SearchSystem6017393645; SearchSystem5219240075; SearchSystem2768350104; SearchSystem6919669052; SearchSystem1986739074; SearchSystem1555480186; SearchSystem3376893470; SearchSystem9530642569; SearchSystem4877790286; SearchSystem8104932799; SearchSystem2313134663; SearchSystem1545325372; SearchSystem7742471461; SearchSystem9092363703; SearchSystem6992236221; SearchSystem3507700306; SearchSystem1129983453; SearchSystem1077927937; SearchSystem2297142691; SearchSystem7813572891; SearchSystem5668754497; SearchSystem6220295595; SearchSystem4157940963; SearchSystem7656671655; SearchSystem2865656762; SearchSystem6520604676; SearchSystem4960161466; .NET CLR 3.0.4506.2152; .NET CLR 3.5.30729)');

    Object.keys(ua).length.should.eql(17);
    ua.type.should.eql('Browser');
    ua.uaFamily.should.eql('IE');
    ua.uaName.should.eql('IE 8.0');
    ua.uaUrl.should.eql('http://en.wikipedia.org/wiki/Internet_Explorer');
    ua.uaCompany.should.eql('Microsoft Corporation.');
    ua.uaCompanyUrl.should.eql('http://www.microsoft.com/');
    ua.uaIcon.should.eql('http://user-agent-string.info/pub/img/ua/msie.png');
    ua.uaInfoUrl.should.eql('http://user-agent-string.info/list-of-ua/browser-detail?browser=IE');
    ua.osFamily.should.eql('Windows');
    ua.osName.should.eql('Windows XP');
    ua.osUrl.should.eql('http://en.wikipedia.org/wiki/Windows_XP');
    ua.osCompany.should.eql('Microsoft Corporation.');
    ua.osCompanyUrl.should.eql('http://www.microsoft.com/');
    ua.osIcon.should.eql('http://user-agent-string.info/pub/img/os/windowsxp.png');
    ua.deviceType.should.eql('Personal computer');
    ua.deviceIcon.should.eql('http://user-agent-string.info/pub/img/device/desktop.png');
    ua.deviceInfoUrl.should.eql('http://user-agent-string.info/list-of-ua/device-detail?device=Personal computer');
  });

  it('parses first 1000 characters of user agent', function() {
    var ua = uasParser.parse('                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 IBrowse');

    Object.keys(ua).length.should.eql(17);
    ua.type.should.eql('Browser');
    ua.uaFamily.should.eql('IBrowse');
    ua.uaName.should.eql('IBrowse');
    ua.uaUrl.should.eql('http://www.ibrowse-dev.net/');
    ua.uaCompany.should.eql('Stefan Burström');
    ua.uaCompanyUrl.should.eql('');
    ua.uaIcon.should.eql('http://user-agent-string.info/pub/img/ua/ibrowse.png');
    ua.uaInfoUrl.should.eql('http://user-agent-string.info/list-of-ua/browser-detail?browser=IBrowse');
    ua.osFamily.should.eql('unknown');
    ua.osName.should.eql('unknown');
    ua.osUrl.should.eql('unknown');
    ua.osCompany.should.eql('unknown');
    ua.osCompanyUrl.should.eql('unknown');
    ua.osIcon.should.eql('http://user-agent-string.info/pub/img/os/unknown.png');
    ua.deviceType.should.eql('Personal computer');
    ua.deviceIcon.should.eql('http://user-agent-string.info/pub/img/device/desktop.png');
    ua.deviceInfoUrl.should.eql('http://user-agent-string.info/list-of-ua/device-detail?device=Personal computer');
  });

  it('ignores everything past the 1000 character of user agent', function() {
    var ua = uasParser.parse('                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        IBrowse');

    Object.keys(ua).length.should.eql(17);
    ua.type.should.eql('unknown');
    ua.uaFamily.should.eql('unknown');
    ua.uaName.should.eql('unknown');
    ua.uaUrl.should.eql('unknown');
    ua.uaCompany.should.eql('unknown');
    ua.uaCompanyUrl.should.eql('unknown');
    ua.uaIcon.should.eql('http://user-agent-string.info/pub/img/ua/unknown.png');
    ua.uaInfoUrl.should.eql('unknown');
    ua.osFamily.should.eql('unknown');
    ua.osName.should.eql('unknown');
    ua.osUrl.should.eql('unknown');
    ua.osCompany.should.eql('unknown');
    ua.osCompanyUrl.should.eql('unknown');
    ua.osIcon.should.eql('http://user-agent-string.info/pub/img/os/unknown.png');
    ua.deviceType.should.eql('Personal computer');
    ua.deviceIcon.should.eql('http://user-agent-string.info/pub/img/device/desktop.png');
    ua.deviceInfoUrl.should.eql('http://user-agent-string.info/list-of-ua/device-detail?device=Personal computer');
  });

  describe('device type', function() {
    it('uses other for robots', function() {
      var ua = uasParser.parse('Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)');

      Object.keys(ua).length.should.eql(17);
      ua.deviceType.should.eql('Other');
      ua.deviceIcon.should.eql('http://user-agent-string.info/pub/img/device/other.png');
      ua.deviceInfoUrl.should.eql('http://user-agent-string.info/list-of-ua/device-detail?device=Other');
    });

    it('performs matching for the device type based on user agent', function() {
      var ua = uasParser.parse('Mozilla/5.0 (iPad; U; CPU OS 4_2_1 like Mac OS X; ja-jp) AppleWebKit/533.17.9 (KHTML, like Gecko) Version/5.0.2 Mobile/8C148 Safari/6533.18.5');

      Object.keys(ua).length.should.eql(17);
      ua.deviceType.should.eql('Tablet');
      ua.deviceIcon.should.eql('http://user-agent-string.info/pub/img/device/tablet.png');
      ua.deviceInfoUrl.should.eql('http://user-agent-string.info/list-of-ua/device-detail?device=Tablet');
    });


    it('uses other for library, validators, useragent anonymizers, and others', function() {
      var agents = [
        'curl/7.30.0',
        'W3C_Validator/1.654',
        'http://Anonymouse.org/ (Unix)',
        'Cyberduck/3.3 (5552)',
      ];

      var ua;
      for(var i = 0; i < agents.length; i++) {
        ua = uasParser.parse(agents[i]);

        ua.deviceType.should.eql('Other');
        ua.deviceIcon.should.eql('http://user-agent-string.info/pub/img/device/other.png');
        ua.deviceInfoUrl.should.eql('http://user-agent-string.info/list-of-ua/device-detail?device=Other');
      }
    });

    it('uses smartphones for mobile and wap browsers', function() {
      var agents = [
        'Mozilla/5.0 (Macintosh; U; Intel Mac OS X 10_5_7; en-us) AppleWebKit/530.17 (KHTML, like Gecko) Version/4.0 Safari/530.17 Skyfire/2.0',
        'Klondike/1.50 (HTTP Win32)',
      ];

      var ua;
      for(var i = 0; i < agents.length; i++) {
        ua = uasParser.parse(agents[i]);

        ua.deviceType.should.eql('Smartphone');
        ua.deviceIcon.should.eql('http://user-agent-string.info/pub/img/device/phone.png');
        ua.deviceInfoUrl.should.eql('http://user-agent-string.info/list-of-ua/device-detail?device=Smartphone');
      }
    });

    it('defaults to personal computer', function() {
      var ua = uasParser.parse('gobbledygook');

      ua.deviceType.should.eql('Personal computer');
      ua.deviceIcon.should.eql('http://user-agent-string.info/pub/img/device/desktop.png');
      ua.deviceInfoUrl.should.eql('http://user-agent-string.info/list-of-ua/device-detail?device=Personal computer');
    });
  });
});


test/updater.js
'use strict';

require('chai').should();

var async = require('async'),
    sinon = require('sinon'),
    Updater = require('../lib/updater');

describe('updater', function() {
  it('only performs 1 check even if multiple updaters are created simultaneously', function(done) {
    this.timeout(15000);

    var checkVersionSpy = sinon.spy(Updater.prototype, 'handleFileStat');

    async.times(5, function(id, callback) {
      var updateInterval = 1 * 24 * 60 * 60 * 1000; // 1 day
      new Updater(updateInterval, function() {
        callback(null);
      });
    }, function() {
      checkVersionSpy.callCount.should.eql(1);
      done();
    });
  });
});


Gruntfile.js
'use strict';

module.exports = function(grunt) {
  grunt.loadNpmTasks('grunt-contrib-jshint');
  grunt.loadNpmTasks('grunt-mocha-test');

  grunt.initConfig({
    jshint: {
      options: {
        jshintrc: '.jshintrc'
      },
      all: [
        'Gruntfile.js',
        'index.js',
        'lib/**/*.js',
        'test/**/*.js',
      ],
    },

    mochaTest: {
      test: {
        options: {
          reporter: 'spec',
        },
        src: ['test/**/*.js']
      },
    },
  });

  grunt.registerTask('updateData', function() {
    var done = this.async();

    var Updater = require('./lib/updater');

    var updateInterval = 1 * 24 * 60 * 60 * 1000; // 1 day
    new Updater(updateInterval, function(error) {
      done(error);
    });
  });

  grunt.registerTask('default', [
    'jshint',
    'updateData',
    'mochaTest',
  ]);
};


index.js
'use strict';

var _ = require('lodash'),
    cache = require('./data/cache'),
    LRU = require('lru-cache'),
    Updater = require('./lib/updater');

var MAX_REASONABLE_LENGTH = 1000;
var updateInterval = 7 * 24 * 60 * 60 * 1000; // 7 days

function finalizeResult(result) {
  if(result.uaInfoUrl !== 'unknown') {
    result.uaInfoUrl = 'http://user-agent-string.info' + result.uaInfoUrl;
  }

  if(result.uaIcon !== 'unknown') {
    result.uaIcon = 'http://user-agent-string.info/pub/img/ua/' + result.uaIcon;
  }

  if(result.osIcon !== 'unknown') {
    result.osIcon = 'http://user-agent-string.info/pub/img/os/' + result.osIcon;
  }

  if(result.deviceIcon !== 'unknown') {
    result.deviceIcon = 'http://user-agent-string.info/pub/img/device/' + result.deviceIcon;
  }

  if(result.deviceInfoUrl !== 'unknown') {
    result.deviceInfoUrl = 'http://user-agent-string.info' + result.deviceInfoUrl;
  }

  return result;
}

exports.parse = function(userAgent) {
  var result = {
    type: 'unknown',
    uaFamily: 'unknown',
    uaName: 'unknown',
    uaUrl: 'unknown',
    uaCompany: 'unknown',
    uaCompanyUrl: 'unknown',
    uaIcon: 'unknown.png',
    uaInfoUrl: 'unknown',
    osFamily: 'unknown',
    osName: 'unknown',
    osUrl: 'unknown',
    osCompany: 'unknown',
    osCompanyUrl: 'unknown',
    osIcon: 'unknown.png',
    deviceType: 'unknown',
    deviceIcon: 'unknown.png',
    deviceInfoUrl: 'unknown',
  };

  // Limit the parsed user agent to the first 1000 characters to prevent regex
  // denial of service attacks.
  if(userAgent.length > MAX_REASONABLE_LENGTH) {
    userAgent = userAgent.substr(0, 1000);
  }

  for(var i = 0; i < cache.robots.order.length; i++) {
    var robotId = cache.robots.order[i];
    var robot = cache.robots[robotId];

    if(robot.userAgent === userAgent) {
      result.type = 'Robot';
      result = _.extend(result, robot.metadata);
      _.extend(result, cache.device['1']);

      return finalizeResult(result);
    }
  }

  var osId;
  for(i = 0; i < cache.browserReg.order.length; i++) {
    var browserRegId = cache.browserReg.order[i];
    var browserReg = cache.browserReg[browserRegId];

    var matches = userAgent.match(browserReg.regexp);
    if(matches) {
      var browser = cache.browser[browserReg.browserId];
      if(browser) {
        result = _.extend(result, browser.metadata);

        var browserType = cache.browserType[browser.typeId];
        if(browserType) {
          result.type = browserType;
        }

        result.uaName = browser.metadata.uaFamily;
        if(matches[1]) {
          result.uaName += ' ' + matches[1];
        }
      }

      osId = cache.browserOs[browserReg.browserId];

      break;
    }
  }

  if(!osId) {
    for(i = 0; i < cache.osReg.order.length; i++) {
      var osRegId = cache.osReg.order[i];
      var osReg = cache.osReg[osRegId];

      if(osReg.regexp.test(userAgent)) {
        osId = osReg.osId;
        break;
      }
    }
  }

  if(osId) {
    var os = cache.os[osId];
    if(os) {
      result = _.extend(result, os);
    }
  }

  var device;
  if(result.type === 'Robot') {
    device = cache.device['1'];
  } else {
    for(i = 0; i < cache.deviceReg.order.length; i++) {
      var deviceRegId = cache.deviceReg.order[i];
      var deviceReg = cache.deviceReg[deviceRegId];

      if(deviceReg.regexp.test(userAgent)) {
        device = cache.device[deviceReg.deviceId];
        break;
      }
    }
  }

  if(!device) {
    if(['Other', 'Library', 'Validator', 'Useragent Anonymizer'].indexOf(result.type) !==  -1) {
      device = cache.device['1'];
    } else if(['Mobile Browser', 'Wap Browser'].indexOf(result.type) !==  -1) {
      device = cache.device['3'];
    } else {
      device = cache.device['2'];
    }
  }

  if(device) {
    result = _.extend(result, device);
  }

  return finalizeResult(result);
};

var lookupCache = LRU({
  max: 5000,
  maxAge: updateInterval,
});

exports.lookup = function(userAgent) {
  var cached = lookupCache.get(userAgent);
  if(!cached) {
    cached = exports.parse(userAgent);
    lookupCache.set(userAgent, cached);
  }

  return cached;
};

function updateData(callback) {
  new Updater(updateInterval, function(error, newCache) {
    if(!error) {
      cache = newCache;
    }

    if(callback) {
      callback(error);
    }
  });
}

updateData();
var interval = setInterval(updateData, updateInterval);
interval.unref();
